### PR TITLE
doc: add more rules in coding guidelines

### PR DESCRIPTION
**Pull request by shiqingg**
This patch adds more rules in coding guidelines.

Tracked-On: #861
Signed-off-by: Shiqing Gao <shiqing.gao@intel.com>

### CODE REVIEW

File: doc/developer-guides/coding_guidelines.rst [hunk 1]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+
+   Non-compliant example::
+
+       uint32_t i;
+       
+       for (i = 0U; i < 5U; i++)
+           printf("count: %d \n", i);
+
+
+ST-14: A space shall be existed between the expression and the control keywords
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: s/be existed/exist/

File: doc/developer-guides/coding_guidelines.rst [hunk 2]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+===============================================================================
+
+A space shall be existed between the expression and the control keywords,
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: s/be existed/exist/

File: doc/developer-guides/coding_guidelines.rst [hunk 3]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+including if, switch, while and for.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
including if, switch, while, and for.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 4]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+   Non-compliant example::
+
+       uint32_t showcase;
+       
+       if(showcase == 0U){
+           showcase = 32U;
+       }
+
+
+ST-15: A space shall be existed after the semi colon in a for expression
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: s/be existed/exist/
s/semi colon/semicolon/

File: doc/developer-guides/coding_guidelines.rst [hunk 5]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+   Non-compliant example::
+
+       uint32_t i;
+       
+       for (i = 0U;i < 5U;i++) {
+           printf("count: %d \n", i);
+       }
+
+
+ST-16: A space shall be existed around binary operators and assignment operators
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: s/be existed/exist/

File: doc/developer-guides/coding_guidelines.rst [hunk 6]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+       uint32_t showcase=32U;
+       
+       showcase=showcase*2U;
+
+
+
+Types
+*****
+
+TY-01: The function return type shall be consistent
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
TY-01: The function return value shall be consistent with the declared return type
```

File: doc/developer-guides/coding_guidelines.rst [hunk 7]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+       int32_t mask = -1;
+       
+       showcase = showcase & mask;
+
+
+TY-03: Mixed-use between Boolean values and integers shall not be allowed
+=========================================================================
+
+Some detailed rules are listed below:
+a) The operands of the arithmetic operation shall be integers;
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: add a blank line before the first item of a list

File: doc/developer-guides/coding_guidelines.rst [hunk 8]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+       uint32_t cond_exp = 64U;
+       
+       uint32_t showcase = showcase_flag ? exp : cond_exp;
+
+
+TY-04: The enum shall not be used for arithmetic operations
+===========================================================
+
+Only the following operations on enum shall be allowed:
+a) enum assignment shall be allowed if the operands of = operation have the same
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: add a blank line before the first item of a list

File: doc/developer-guides/coding_guidelines.rst [hunk 9]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+enum type;
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: indent a list continuation line

File: doc/developer-guides/coding_guidelines.rst [hunk 10]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+b) enum comparison shall be allowed, including the operators ==, !=, >, <, >=
+and <=.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: indent

File: doc/developer-guides/coding_guidelines.rst [hunk 11]
```diff
@@ -451,11 +735,299 @@ Compliant example::
...
+       uint32_t showcase = showcase_flag ? exp : cond_exp;
+
+
+TY-04: The enum shall not be used for arithmetic operations
+===========================================================
+
+Only the following operations on enum shall be allowed:
+a) enum assignment shall be allowed if the operands of = operation have the same
+enum type;
+b) enum comparison shall be allowed, including the operators ==, !=, >, <, >=
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
b) enum comparison shall be allowed, including the operators ==, !=, >, <, >=,
```

File: doc/developer-guides/coding_guidelines.rst [hunk 12]
```diff
@@ -475,3 +1047,241 @@ Compliant example::
...
+       };
+       
+       struct struct_showcase showcase = {32U, -1};
+
+
+
+Identifiers
+***********
+
+ID-01: A parameter name shall not be the same as the name of struct, union or enum
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
ID-01: A parameter name shall not be the same as the name of struct, union, or enum
```

File: doc/developer-guides/coding_guidelines.rst [hunk 13]
```diff
@@ -475,3 +1047,241 @@ Compliant example::
...
+           char *str_destination;
+       };
+       
+       void func_showcase(uint32_t showcase)
+       {
+           /* main body */
+       }
+
+
+ID-02: A member name shall not be the same as the name of struct, union or enum
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
ID-02: A member name shall not be the same as the name of struct, union, or enum
```

File: doc/developer-guides/coding_guidelines.rst [hunk 14]
```diff
@@ -475,3 +1047,241 @@ Compliant example::
...
+       };
+       
+       struct showcase_2
+       {
+           uint32_t showcase_1;
+           uint32_t showcase_2;
+       };
+
+
+ID-03: A global variable name shall not be the same as the name of struct, union or enum
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
ID-03: A global variable name shall not be the same as the name of struct, union, or enum
```

File: doc/developer-guides/coding_guidelines.rst [hunk 15]
```diff
@@ -475,3 +1047,241 @@ Compliant example::
...
+       
+           showcase = 32U;
+       }
+
+
+ID-05: The function name shall be unique
+========================================
+
+The function name shall not be the same as the name of struct, union, enum,
+variable or other functions.
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]:
```suggestion
variable, or other functions.
```

File: doc/developer-guides/coding_guidelines.rst (hunk 12 above)
Comment (dbkinder) [reply doc/developer-guides/coding_guidelines.rst]: parameter names can't conflict with variable names or function names either, right?

File: doc/developer-guides/coding_guidelines.rst [hunk 16]
```diff
@@ -7,6 +7,36 @@ Coding Guidelines
    :local:
 
 
+Preprocessor
+************
+
+PP-01: ## or # operators shall be used with restrictions
+========================================================
+
+## or # operators shall be used carefully. The following cases shall not be
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: "shall be used carefully" is not a clear requirement.  What does this mean?

**Comment by shiqingg** [comment]
@dbkinder
David, thanks for your review. 
I have pushed the new version per your suggestion.

**Pull request merged**